proof/refine/Retype_R.thy

```diff
@@ -1607,1195 +1350,904 @@ next
+   (* avoid unfolding word_bits_def *)
+   apply (simp add: of_bl_length[where 'a=machine_word_len, folded word_bits_def])
+  apply (clarsimp simp: image_def cte_map_def shiftl_t2n')
+  apply (rule_tac x="drop (word_bits - n) (to_bl xa)" in exI)
+   (* avoid revealing word size via LENGTH *)
+  supply word_bl_Rep'[simp del]
+         word_bl_Rep'[where 'a=machine_word_len, folded word_bits_def, simp]
+  apply (simp add: of_drop_to_bl word_bits_size less_mask_eq)
+  done
 
-lemma valid_n:
-  "valid_mdb_ctes n"
-  by (simp add: valid_mdb_ctes_def dlist_n no_0_n mdb_chain_0_n
-                valid_badges_n caps_contained_n untyped_mdb_n
-                untyped_inc_n mdb_chunked_n valid_nullcaps_n ut_rev_n
-                class_links_n irq_control_n dist_z_n
-                reply_masters_rvk_fb_n)
+lemma (in Retype_R_2) corres_retype:
+  assumes    not_zero: "n \<noteq> 0"
+  and         aligned: "is_aligned ptr (objBitsKO ko + gbits)"
+  and    obj_bits_api: "obj_bits_api (APIType_map2 ty) us = objBitsKO ko + gbits"
+  and              tp: "APIType_map2 ty \<in> no_gs_types"
+  and              ko: "makeObjectKO dev d ty = Some ko"
+  and             orr: "obj_bits_api (APIType_map2 ty) us \<le> sz \<Longrightarrow>
+                        obj_relation_retype (default_object (APIType_map2 ty) dev us d) ko"
+  and           cover: "range_cover ptr sz (obj_bits_api (APIType_map2 ty) us) n"
+  shows
+    "corres (=)
+       (\<lambda>s. valid_pspace s \<and> pspace_no_overlap_range_cover ptr sz s \<and> valid_mdb s \<and> valid_list s)
+       (\<lambda>s. pspace_aligned' s \<and> pspace_distinct' s \<and> pspace_no_overlap' ptr sz s
+            \<and> (\<exists>val. ko = injectKO val)
+            \<and> (ty = Inr (APIObjectType TCBObject) \<longrightarrow> d = ksCurDomain s))
+       (retype_region ptr n us (APIType_map2 ty) dev) (createObjects ptr n ko gbits)"
+  apply (rule corres_guard_imp)
+    apply (rule_tac F = "(\<exists>val. ko = injectKO val)" in corres_gen_asm2)
+    apply (erule exE)
+    apply (rule corres_rel_imp)
+    apply (rule corres_retype'[where g=id and ty=ty and sz = sz,OF not_zero aligned _ _ _ ko,
+                                     simplified update_gs_id[OF tp] modify_id_return,simplified])
+        using assms
+        apply (simp_all add: objBits_def no_gs_types_CapTableObject)
+  apply auto
+  done
 
-end
+lemmas gen_object_splits =
+  apiobject_type.split_asm
+  sum.split_asm kernel_object.split_asm
 
-definition
-  caps_no_overlap'' :: "machine_word \<Rightarrow> nat \<Rightarrow> kernel_state \<Rightarrow> bool"
-where
- "caps_no_overlap'' ptr sz s \<equiv> \<forall>cte \<in> ran (ctes_of s).
-               untypedRange (cteCap cte) \<inter> {ptr .. (ptr && ~~ mask sz) + 2 ^ sz - 1} \<noteq> {}
-               \<longrightarrow> {ptr .. (ptr && ~~ mask sz) + 2 ^ sz - 1} \<subseteq> untypedRange (cteCap cte)"
+declare hoare_in_monad_post[wp del]
+declare univ_get_wp[wp del]
```

Comment:
Not relevant to this PR at all but these declares should probably have a FIXME on them to be cleaned up.

Reply:
~~I'll try a cleanup pass today~~ actually I have no idea, will go with the FIXME